File: GonePostal/GPSetController.m
//
//  GPSetController.m
//  GonePostalX
//
//  Created by Travis Gruber on 2/13/13.
//  Copyright (c) 2013 Travis Gruber. All rights reserved.
//

#import "GPSetController.h"

@interface GPSetController ()

@end

@implementation GPSetController

- (id)initWithWindow:(NSWindow *)window
{
    self = [super initWithWindow:window];
    if (self) {
        NSSortDescriptor *setSort = [[NSSortDescriptor alloc] initWithKey:@"name" ascending:YES];
        self.gpCatalogSetsSortDescriptors = @[setSort];
        
        NSSortDescriptor *gpCatalogSort = [[NSSortDescriptor alloc] initWithKey:@"gp_catalog_number" ascending:YES];
        self.gpCatalogEntriesSortDescriptors = @[gpCatalogSort];
        
        NSSortDescriptor *countrySort = [[NSSortDescriptor alloc] initWithKey:@"country_name" ascending:YES];
        self.countrySortDescriptors = @[countrySort];
        
        NSSortDescriptor *sectionSort = [[NSSortDescriptor alloc] initWithKey:@"group_name" ascending:YES];
        self.sectionSortDescriptors = @[sectionSort];
    }
    
    return self;
}

- (void)windowDidLoad
{
    [super windowDidLoad];
}

- (NSString *)windowTitleForDocumentDisplayName:(NSString *)displayName {
    return @"Catalog Sets";
}

- (IBAction)addSet:(id)sender {
    [self.gpCatalogSetsController insert:sender];
    
    NSError * error;
    if (![self.managedObjectContext save:&error]) {
        NSAlert * errSheet = [NSAlert alertWithError:error];
        [errSheet beginSheetModalForWindow:self.window modalDelegate:nil didEndSelector:nil contextInfo:nil];
        [self.managedObjectContext undo];
    }
}

- (IBAction)removeSet:(id)sender {
    [self.gpCatalogSetsController remove:sender];
    
    NSError * error;
    if (![self.managedObjectContext save:&error]) {
        NSAlert * errSheet = [NSAlert alertWithError:error];
        [errSheet beginSheetModalForWindow:self.window modalDelegate:nil didEndSelector:nil contextInfo:nil];
        [self.managedObjectContext undo];
    }
}

- (IBAction)removeCatalogEntries:(id)sender {
    [self.gpCatalogsController removeObjects:self.gpCatalogsController.selectedObjects];
    
    NSError * error;
    if (![self.managedObjectContext save:&error]) {
        NSAlert * errSheet = [NSAlert alertWithError:error];
        [errSheet beginSheetModalForWindow:self.window modalDelegate:nil didEndSelector:nil contextInfo:nil];
        [self.managedObjectContext undo];
    }
}

@end

File: GonePostal/GPSetController.swift
//
//  GPSetController.swift
//  GonePostal
//
//  Created by Travis Gruber on 11/17/16.
//  Copyright © 2016 Travis Gruber. All rights reserved.
//

import Cocoa

internal class GPSetController: NSWindowController {
    
    let gpCatalogSetsSortDescriptors = [NSSortDescriptor(key:"name", ascending: true)] as NSArray
    let gpCatalogEntriesSortDescriptors = [NSSortDescriptor(key:"gp_catalog_number", ascending: true)] as NSArray
    let cachetsSortDescriptors = [NSSortDescriptor(key:"gp_cachet_number", ascending: true)] as NSArray
    let countrySortDescriptors = [NSSortDescriptor(key:"country_name", ascending: true)] as NSArray
    let sectionSortDescriptors = [NSSortDescriptor(key:"group_name", ascending: true)] as NSArray
    
    var managedObjectContext: NSManagedObjectContext
    
    @IBOutlet var gpCatalogSetsController: NSArrayController!
    @IBOutlet var gpCatalogsController: NSArrayController!
    @IBOutlet var cachetsController: NSArrayController!
    
    override var windowNibName: String! {
        return "GPSetController"
    }
    
    override init(window: NSWindow?) {
        let gpDocument = NSDocumentController.shared.currentDocument as! GPDocument
        managedObjectContext = gpDocument.managedObjectContext!
        
        super.init(window:window)
    }
    
    required init?(coder: NSCoder) {
        fatalError("init(coder:) has not been implemented. use init(gpCatalog:)")
    }
    
    @IBAction func addSet(_ sender: AnyObject) {
        gpCatalogSetsController.insert(sender)
        save()
    }
    
    @IBAction func removeSet(_ sender: AnyObject) {
        gpCatalogSetsController.remove(sender)
        save()
    }
    
    @IBAction func addCatalogEntries(_ sender: AnyObject) {
        
    }
    
    @IBAction func removeCatalogEntries(_ sender: AnyObject) {
        gpCatalogsController.remove(contentsOf: self.gpCatalogsController.selectedObjects)
        save()
    }
    
    @IBAction func addCachets(_ sender: AnyObject) {
        if let selection = self.gpCatalogSetsController.selectedObjects {
            if (!selection.isEmpty) {
                let set = selection[0] as! GPCatalogSet
                let controller = GPCachetEditor(gpCatalogSet: set)
                self.document!.addWindowController(controller)
                controller.window!.makeKeyAndOrderFront(sender)
            }
        }
    }
    
    @IBAction func removeCachets(_ sender: AnyObject) {
        cachetsController.remove(contentsOf: self.cachetsController.selectedObjects)
        save()
    }
    
    func save() {
        do {
            try self.managedObjectContext.save()
        } catch {
            let alert = NSAlert(error: error)
            alert.beginSheetModal(for: self.window!, completionHandler: nil)
            self.managedObjectContext.rollback()
        }
    }
}
